openhands/runtime/impl/runloop/runloop_runtime.py

```diff
@@ -0,0 +1,259 @@
+        self._stop_event.set()
+        self.log_stream_thread.join(timeout)
+
+
+class RunloopRuntime(EventStreamRuntime):
+    """The RunloopRuntime class is an EventStreamRuntime that utilizes Runloop Devbox as a runtime environment."""
+
+    _sandbox_port: int = 4444
+
+    def __init__(
+        self,
+        config: AppConfig,
+        event_stream: EventStream,
+        sid: str = 'default',
+        plugins: list[PluginRequirement] | None = None,
+        env_vars: dict[str, str] | None = None,
+        status_message_callback: Callable | None = None,
+        attach_to_existing: bool = False,
+    ):
+        assert config.runloop_api_key is not None, 'Runloop API key is required'
+        self.devbox: DevboxView | None = None
+        self.config = config
+        self.runloop_api_client = Runloop(
+            bearer_token=config.runloop_api_key,
+            base_url='https://api.runloop.pro',
+        )
+        self.session = requests.Session()
+        self.status_message_callback = status_message_callback
+        self.container_name = self.container_name_prefix + sid
+        self.action_semaphore = threading.Semaphore(1)  # Ensure one action at a time
+        self.init_base_runtime(
+            config,
+            event_stream,
+            sid,
+            plugins,
+            env_vars,
+            status_message_callback,
+            attach_to_existing,
+        )
+
+    @tenacity.retry(
+        stop=tenacity.stop_after_attempt(90),
+        wait=tenacity.wait_fixed(0.75),
+    )
+    def _wait_for_devbox(self):
+        """Pull devbox status until it is running"""
+        if self.devbox is None:
+            raise ConnectionRefusedError('Devbox has not been created')
+
+        if self.devbox.status == 'running':
+            return
+
+        devbox = self.runloop_api_client.devboxes.retrieve(id=self.devbox.id)
+        if devbox.status != 'running':
+            raise ConnectionRefusedError('Devbox is not running')
+
+        # Devbox is connected and running
+        self.devbox = devbox
+
+        logging.debug(f'devbox.id={self.devbox.id} is running')
+
+    async def connect(self):
+        self.send_status_message('STATUS$STARTING_RUNTIME')
+
+        # Note: Runloop connect
+        sandbox_workspace_dir = (
+            self.config.workspace_mount_path_in_sandbox
+        )  # e.g. /workspace
+        plugins = self.plugins
+        plugin_arg = ''
+        if plugins is not None and len(plugins) > 0:
+            plugin_arg = f'--plugins {" ".join([plugin.name for plugin in plugins])} '
+        if self.config.sandbox.browsergym_eval_env is not None:
+            browsergym_arg = (
+                f'--browsergym-eval-env {self.config.sandbox.browsergym_eval_env}'
+            )
+        else:
+            browsergym_arg = ''
+
+        # Copied from EventstreamRuntime
+        start_command = (
+            f'/openhands/micromamba/bin/micromamba run -n openhands '
```

Comment:
I recently factored this out to https://github.com/All-Hands-AI/OpenHands/blob/d50425865a08387e4cc6548242b6b4fd965686ec/openhands/runtime/utils/command.py#L10 if you're able to reuse that!

Reply:
done!